js/client/modules/@arangodb/testsuites/endpoints.js

```diff
@@ -59,69 +59,187 @@ function endpoints (options) {
     let testName = 'endpoint-' + endpointName;
-    results[testName] = (function () {
-      let endpoint = endpoints[endpointName]();
-      if (endpoint === undefined || options.cluster || options.skipEndpoints) {
-        return {
-          failed: 0,
-          status: true,
-          skipped: true
-        };
-      } else {
-        let instanceInfo = pu.startInstance('tcp', Object.assign(options, {useReconnect: true}), {
-          'server.endpoint': endpoint
-        }, testName);
+    let testCase = endpoints[endpointName];
+    if (options.cluster || options.skipEndpoints || (testCase.skip && testCase.skip())) {
+      return {
+        failed: 0,
+        status: true,
+        skipped: true
+      };
+    }
 
-        if (instanceInfo === false) {
-          results.failed += 1;
-          return {
-            failed: 1,
-            status: false,
-            message: 'failed to start server!'
-          };
-        }
+    let serverArgs = testCase.serverArgs();
+    if (typeof serverArgs === 'string') {
+      serverArgs = { 'server.endpoint': serverArgs };
+    }
 
-        let oneTestResult = tu.runInArangosh(options, instanceInfo, testPaths.endpoints[0]);
+    let instanceInfo = pu.startInstance('tcp', Object.assign(options, {useReconnect: true}), serverArgs, testName);
 
-        print(CYAN + 'Shutting down...' + RESET);
-        // mop: mehhh...when launched with a socket we can't use download :S
-        oneTestResult['shutdown'] = pu.shutdownInstance(instanceInfo, Object.assign(options, {useKillExternal: true}));
-        print(CYAN + 'done.' + RESET);
+    if (instanceInfo === false) {
+      results.failed += 1;
+      return {
+        failed: 1,
+        status: false,
+        message: 'failed to start server!'
+      };
+    }
 
-        if (!oneTestResult.status || !oneTestResult.shutdown) {
+    const specFile = testPaths.endpoints[0];
+    
+    testCase.shellTests.forEach(function(testCase) {
+      let old = instanceInfo.endpoint;
+      let shellEndpoint = testCase.endpoint(serverArgs['server.endpoint']);
+      instanceInfo.endpoint = shellEndpoint;
+      try {
+        let result = tu.runInArangosh(options, instanceInfo, specFile, { 'server.connection-timeout': 2, 'server.request-timeout': 2 });
+        let success = result.status === testCase.success;
+        results[endpointName + '-' + testCase.name] = { status: success }; 
+        if (!success) {
+          // arangosh or the test returned an error
+          results[endpointName + '-' + testCase.name].message = result.message;
           results.failed += 1;
-          results.shutdown = false;
-        } else {
-          pu.cleanupLastDirectory(options);
         }
-        return oneTestResult;
+      } finally {
+        instanceInfo.endpoint = old;
       }
-    }());
+    });
+
+    print(CYAN + 'Shutting down...' + RESET);
+    let shutdown = pu.shutdownInstance(instanceInfo, Object.assign(options, {useKillExternal: true}));
```

Comment:
shouldn't this arangosh (that runs this test) still be able to connect to the server, and hence be able to use the regular shutdown command?

Reply:
I don't understand the question.
We are starting a few server instances in the endpoints test, because the server instances use different options.
For each server instance, we run mutliple arangosh tests, e.g.
```
- server instance1
  - arangosh1  
  - arangosh2  
  - arangosh3  
- server instance2
  - arangosh4  
  - arangosh5  
  - arangosh6  
...
The shutdown command in the test was actually there before, and I didn't modify it. 
What is the exact problem with it?